dcap/types/src/quote3.rs

```diff
@@ -0,0 +1,457 @@
+// Copyright (c) 2022 The MobileCoin Foundation
+// ! This module provides types related to Quote v3
+
+use displaydoc::Display;
+use mc_sgx_dcap_sys_types::quote3_error_t;
+
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+/// An enumeration of errors which occur when using QuoteLib-related methods.
+///
+/// These errors correspond to error elements of
+/// [`quote3_error_t`](mc_sgx_dcap_sys_types::quote3_error_t).
+#[derive(Copy, Clone, Debug, Display, Eq, Hash, Ord, PartialEq, PartialOrd)]
+#[cfg_attr(feature = "serde", derive(Deserialize, Serialize))]
+#[non_exhaustive]
+#[repr(u32)]
+pub enum Error {
+    /// An unexpected internal error occurred
+    Unexpected = quote3_error_t::SGX_QL_ERROR_UNEXPECTED.0,
+    /// One of the parameters passed to an SGX FFI method was invalid
+    InvalidParameter = quote3_error_t::SGX_QL_ERROR_INVALID_PARAMETER.0,
+    /// Not enough memory is available to complete this operation
+    OutOfMemory = quote3_error_t::SGX_QL_ERROR_OUT_OF_MEMORY.0,
+    /// Expected ECDSA_ID does not match the value stored in the ECDSA Blob
+    EcdsaIdMismatch = quote3_error_t::SGX_QL_ERROR_ECDSA_ID_MISMATCH.0,
+    /// The ECDSA blob pathname is too large
+    PathnameBufferOverflow = quote3_error_t::SGX_QL_PATHNAME_BUFFER_OVERFLOW_ERROR.0,
+    /// Error accessing ECDSA blob
+    FileAccess = quote3_error_t::SGX_QL_FILE_ACCESS_ERROR.0,
+    /// Cached ECDSA key is invalid
+    StoredKey = quote3_error_t::SGX_QL_ERROR_STORED_KEY.0,
+    /// Cached ECDSA key does not match requested key
+    PubKeyIdMismatch = quote3_error_t::SGX_QL_ERROR_PUB_KEY_ID_MISMATCH.0,
+    /// PCE use the incorrect signature scheme
+    InvalidPceSigScheme = quote3_error_t::SGX_QL_ERROR_INVALID_PCE_SIG_SCHEME.0,
+    /// There is a problem with the attestation key blob
+    AttestationKeyBlob = quote3_error_t::SGX_QL_ATT_KEY_BLOB_ERROR.0,
+    /// Unsupported attestation key ID
+    UnsupportedAttKeyId = quote3_error_t::SGX_QL_UNSUPPORTED_ATT_KEY_ID.0,
+    /// Unsupported enclave loading policy
+    UnsupportedLoadingPolicy = quote3_error_t::SGX_QL_UNSUPPORTED_LOADING_POLICY.0,
+    /// Unable to load the QE enclave
+    InterfaceUnavailable = quote3_error_t::SGX_QL_INTERFACE_UNAVAILABLE.0,
+    /// Unable to find the platform library with the dependent APIs (not fatal)
+    PlatformLibUnavailable = quote3_error_t::SGX_QL_PLATFORM_LIB_UNAVAILABLE.0,
+    /// The attestation key doesn't exist or has not been certified
+    AttestationKeyNotInitialized = quote3_error_t::SGX_QL_ATT_KEY_NOT_INITIALIZED.0,
+    /// The certification data retrieved from the platform library is invalid
+    InvalidCertDataInAttestationKey = quote3_error_t::SGX_QL_ATT_KEY_CERT_DATA_INVALID.0,
```

Comment:
Hmm, I think I'm OK with us just using `Cert` just using everywhere, because I feel like I never question that "cert" means "certificate". Intel's docs are overly verbose and call it "certification data," which I assume is out of some misguided sense of pomposity...